### ISSUE TITLE
GNUFrame is still needed?

### ISSUE THREAD

**Issue by audetto** [sh95014/AppleWin#37]
Hi,

I did introduce `GNUFrame` to ease the native mac os build.

https://github.com/audetto/AppleWin/commit/f03c09db764578329e736f141e51b25626e7f35f

I had a quick look at this repo and I am not sure it is (ever was) still needed?

Can I merge CommonFrame and GNUFrame back together?

**Comment by audetto** [comment]
Actually forget this. It is still needed by `MarianiFrame`.
No prob.

**Comment by audetto** [comment]
I wanted to add an argument to the constructor of CommonFrame, but I am getting confused with multiple / virtual inheritance... let me revisit it.

**Comment by sh95014** [comment]
@audetto I'm not building `GNUFrame` and `SDLRendererFrame` at all.

Anyway, I would suggest revisiting the entire hierarchy of "frames". Maybe like this:

```mermaid
classDiagram
    FrameBase <|-- UnixFrame
    UnixFrame <|-- MarianiFrame
    UnixFrame <|-- LinuxFrame
    LinuxFrame <|-- SDLFrame
    class UnixFrame {
        (currently LinuxFrame)
        Lifecycle
        Execute
        FrameBuffer
    }
    class LinuxFrame {
        Directories
    }
    class SDLFrame {
        Bitmaps
        Key/mouse/drag/drop handling
        Window title, icon
    }
    class MarianiFrame {
        Directories
        Bitmaps
        Key/mouse/drag/drop handling
    }
```

I can try refactoring it as above and sending you a PR if you like.

**Comment by audetto** [comment]
The only thing I need to be careful about is Lifecycle and Execute as I still want to handle libretro and the ncurses frontend (which could be improved anyway).

Are you able to tell me exactly which methods (from SDL Frame) you use?

**Comment by sh95014** [comment]
> Are you able to tell me exactly which methods (from SDL Frame) you use?

Yup. Here's the diff that reparents `MarianiFrame` on `CommonFrame` instead. Note that I couldn't figure out how to get `SDLRendererFrame` to invoke the `CommonFrame(options)` constructor so I had to insert a dummy default constructor `CommonFrame()` that *doesn't* read the program options.

https://github.com/sh95014/AppleWin/commit/e8692863b5c44afbf872fe281af307ccb9eda9ed

(This change would bring the `#ifdef MARIANI`s down to just eight, which is pretty exciting...)

**Comment by audetto** [comment]
Thank you. I will have a look. Your issue sounds very much the same as mine, so I hope we can fix it.

**Comment by sh95014** [comment]
https://github.com/sh95014/AppleWin/tree/frame-refactor-2023 removes `MarianiFrame`'s reliance on `SDLFrame`, and it hasn't been reliant on `GNUFrame` anyway, so you might want to use https://github.com/audetto/AppleWin/pull/124 to just get rid of `GNUFrame` entirely?

**Comment by audetto** [comment]
I have merged https://github.com/audetto/AppleWin/pull/124

I guess I can sort out `GNUFrame` now.

**Comment by sh95014** [comment]
With your latest changes, I was able to remove all downstream `#ifdef`s except [one](https://github.com/AppleWin/AppleWin/issues/1265), so you really got AppleWin building pretty cleanly for Linux and macOS at this point. Thanks!

**Comment by sh95014** [comment]
@audetto `MarianiFrame` now inherits from `common2::CommonFrame`. Any reason to keep this issue open?
